plugins/catalog-graph/src/components/EntityRelationsGraph/useEntityStore.ts

```diff
@@ -0,0 +1,120 @@
+/*
+ * Copyright 2021 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import { Entity, parseEntityRef } from '@backstage/catalog-model';
+import { useApi } from '@backstage/core-plugin-api';
+import { catalogApiRef } from '@backstage/plugin-catalog-react';
+import limiterFactory from 'p-limit';
+import { Dispatch, useCallback, useRef, useState } from 'react';
+import { useAsyncFn } from 'react-use';
+
+// TODO: This is a good use case for a graphql API, once it is available in the
+// future.
+
+/**
+ * Ensures that a set of requested entities is loaded.
+ */
+export function useEntityStore(): {
+  entities: { [key: string]: Entity };
+  loading: boolean;
+  error?: Error;
+  requestEntities: Dispatch<string[]>;
+} {
+  const catalogClient = useApi(catalogApiRef);
+  const state = useRef({
+    requestedEntities: new Set<string>(),
+    outstandingEntities: new Map<string, Promise<Entity | undefined>>(),
+    cachedEntities: new Map<string, Entity>(),
+  });
+  const [entities, setEntities] = useState<{
+    [key: string]: Entity;
+  }>({});
+
+  const updateEntities = useCallback(() => {
+    const { cachedEntities, requestedEntities } = state.current;
+    const filteredEntities: { [key: string]: Entity } = {};
+    requestedEntities.forEach(entityRef => {
+      const entity = cachedEntities.get(entityRef);
+
+      if (entity) {
+        filteredEntities[entityRef] = entity;
+      }
+    });
+    setEntities(filteredEntities);
+  }, [state, setEntities]);
+
+  const [asyncState, fetch] = useAsyncFn(async () => {
+    const limiter = limiterFactory(10);
```

Comment:
this gives 10 concurrent per each fetch, normally ok but if you wanted to you could move the limiter factory call further up the stack - potentially all the way up at the root of this file, if you wanted to